cmd/main.go

```diff
@@ -0,0 +1,56 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	appcontroller "github.com/argoproj/argo-cd/cmd/argocd-application-controller/commands"
+	dex "github.com/argoproj/argo-cd/cmd/argocd-dex/commands"
+	reposerver "github.com/argoproj/argo-cd/cmd/argocd-repo-server/commands"
+	apiserver "github.com/argoproj/argo-cd/cmd/argocd-server/commands"
+	util "github.com/argoproj/argo-cd/cmd/argocd-util/commands"
+	cli "github.com/argoproj/argo-cd/cmd/argocd/commands"
+	cmdutil "github.com/argoproj/argo-cd/cmd/util"
+)
+
+const (
+	binaryNameEnv = "ARGOCD_BINARY_NAME"
+)
+
+func main() {
+	var command *cobra.Command
+
+	binaryName := filepath.Base(os.Args[0])
+	if val := os.Getenv(binaryNameEnv); val != "" {
+		binaryName = val
+	}
+	switch binaryName {
+	case "argocd", "argocd-linux-amd64", "argocd-darwin-amd64", "argocd-windows-amd64.exe":
+		command = cli.NewCommand()
+	case "argocd-util", "argocd-util-linux-amd64", "argocd-util-darwin-amd64", "argocd-util-windows-amd64.exe":
+		command = util.NewCommand()
+	case "argocd-server":
+		command = apiserver.NewCommand()
+	case "argocd-application-controller":
+		command = appcontroller.NewCommand()
+	case "argocd-repo-server":
+		command = reposerver.NewCommand()
+	case "argocd-dex":
+		command = dex.NewCommand()
+	default:
+		if suggestionsString := cmdutil.FindSuggestions(binaryName); suggestionsString != "" {
```

Comment:
I guess I'm missing something, I cannot understand why this is required. I've tried to comment this line and getting suggestions:

```
argocd cer                                                                                                                                                                  (⎈ |minikube:argocd-e2e)
Error: unknown command "cer" for "argocd"

Did you mean this?
	cert

```

Reply:
I added this change based on discussion in contributor meeting. This part provides suggestion for unknown binary names. So, that if someone for example create a different symlink name for argocd-util, we can try to catch it intelligently. 

```
$ ln -s ./dist/argocd argocd-util-v1

$ ./argocd-util-v1
unknown command "argocd-util-v1" for

Did you mean this?
	argocd-util
```
 What you pasted above is suggestion for subcommand, which cobra does it internally.

Reply:
Ah, got it. Thank you for the explanation! The problem is that if binary is named unexpectedly we would not show any suggestion to the users. What if we try to "guess" the binary name by subcommand: check if either `argocd` or `argocd-util` has given subcommand, use one that has it or show error message if we cannot guess:


```golang
	default:
		if len(os.Args[1:]) > 0 {
			// try to guess between argocd and argocd-util by matching sub command
			for _, cmd := range []*cobra.Command{cli.NewCommand(), util.NewCommand()} {
				if _, _, err := cmd.Find(os.Args[1:]); err == nil {
					command = cmd
					break
				}
			}
		}

		if command == nil {
			fmt.Printf(`Unknown binary name '%s'.
Use '%s' environment variable to specify required binary name (possible values 'argocd' or 'argocd-util').`,
				binaryName, binaryNameEnv)
			os.Exit(1)
		}
```